src/MongoDB.Driver/Linq/Processors/ExpressionValidator.cs

```diff
@@ -0,0 +1,122 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace MongoDB.Driver.Linq.Processors
+{
+    /// <summary>
+    /// The expression visitor to validate a LINQ expression.
+    /// </summary>
+    internal sealed class ExpressionValidator : ExpressionVisitor
+    {
+        private readonly List<string> _topLevelInParameters = new List<string>();
+        private List<string> _inParametersScope = new List<string>();
+
+        /// <summary>
+        /// Validates the input expression.
+        /// </summary>
+        /// <param name="expression">The expression.</param>
+        public static void Validate(Expression expression)
+        {
+            var validator = new ExpressionValidator();
+
+            var lambdaExpression = expression as LambdaExpression;
+            if (lambdaExpression?.Parameters != null && lambdaExpression.Parameters.Any())
+            {
+                validator._topLevelInParameters.AddRange(lambdaExpression.Parameters.Select(c => c.Name));
+            }
+
+            validator.Visit(expression);
+        }
+
+        /// <inheritdoc />
+        protected override Expression VisitMethodCall(MethodCallExpression node)
```

Comment:
The target of this ticket is to check that we don't use unreachable parameters from other(higher) expression levels. As I know, the only one way to walk through the expression levels is by calling methods.
>Couldn't the invalid parameter occur in a unary or binary expression as well?

I guess no, but we can check any situation you mean :)

Reply:
>That's not a MethodCallExpression.
yep, but I'm checking all used expression fields inside of `MethodCallExpression`. So, since invalid1 + invalid2 are placed in the `MethodCallExpression`, it doesn't matter which exact expression type they have.

`Expression<Func<TestObject, bool>> expr = (a) => a.Collection1.Any(b => b.Value1 == 2 && a.Value1 == 3);`

this expression throws the error.